lib/api_watch_diff.ml

```diff
@@ -81,69 +96,123 @@ let diff_value ~typing_env ~val_name ~reference ~current =
-                (Value (val_name, Modified { ref = ref_vd; current = curr_vd }))
-          ))
-    ref_values curr_values
-  |> FieldMap.bindings |> List.map snd
-
-let diff_interface ~reference ~current =
-  let value_diffs = compare_values ~reference ~current in
-  if value_diffs = [] then
-    let typing_env = Env.empty in
-    let coercion1 () =
-      Includemod.signatures typing_env ~mark:Mark_both reference current
-    in
-    let coercion2 () =
-      Includemod.signatures typing_env ~mark:Mark_both current reference
-    in
-    match (coercion1 (), coercion2 ()) with
-    | Tcoerce_none, Tcoerce_none -> []
-    | _, _ -> [ Any ]
-    | exception Includemod.Error _ -> [ Any ]
-  else value_diffs
-
-let to_text_diff (diff_result : diff list) : Diffutils.Diff.t =
-  let vd_to_string name vd =
-    let buf = Buffer.create 256 in
-    let formatter = Format.formatter_of_buffer buf in
-    Printtyp.value_description (Ident.create_local name) formatter vd;
-    Format.pp_print_flush formatter ();
-    Buffer.contents buf
+  let ref_items = extract_items reference in
+  let curr_items = extract_items current in
+  let item_changes =
+    Sig_item_map.merge
+      (fun (item_type, name) ref_opt curr_opt ->
+        match (item_type, ref_opt, curr_opt) with
+        | Value_item, ref_opt, curr_opt -> (
+            match (ref_opt, curr_opt) with
+            | None, None -> None
+            | None, Some (Val curr_vd) ->
+                Some (Value { name; change = Added curr_vd })
+            | Some (Val ref_vd), None ->
+                Some (Value { name; change = Removed ref_vd })
+            | Some (Val ref_vd), Some (Val curr_vd) -> (
+                let value_differs =
+                  diff_value ~typing_env:env ~val_name:name ~reference:ref_vd
+                    ~current:curr_vd
+                in
+                match value_differs with
+                | None -> None
+                | Some _ ->
+                    Some
+                      (Value
+                         {
+                           name;
+                           change =
+                             Modified { ref_ = ref_vd; current = curr_vd };
+                         }))
+            | _ -> None)
+        | Module_item, ref_opt, curr_opt -> (
+            match (ref_opt, curr_opt) with
+            | None, None -> None
+            | None, Some (Mod _curr_md) ->
+                Some (Module { module_name = name; changes = Unsupported })
+            | Some (Mod _ref_md), None ->
+                Some (Module { module_name = name; changes = Unsupported })
+            | Some (Mod ref_md), Some (Mod curr_md) -> (
+                match (ref_md.md_type, curr_md.md_type) with
+                | Mty_signature ref_submod, Mty_signature curr_submod ->
+                    diff_items name ~reference:ref_submod ~current:curr_submod
+                | _, _ ->
+                    Some (Module { module_name = name; changes = Unsupported }))
```

Comment:
This is incorrect, you can't state that there will necessarily be a change there.

It's okay to stick to simple module declaration as you did above but we should default to calling some `Includemod.signatures` (probably `Includemod.modtypes`) like function here to detect changes when we don't know how to diff those.